### PR TITLE
chore: csdk macros3

### PR DESCRIPTION
**Pull request by SwenSchaeferjohann**
stacked on top of #2020, merge 2020 first

* rent impl for c-PDA.

reuses most of the rent impl of c-token, however we always transfer rent to the program's rent_sponsor pda, even if a different fee_payer paid for rent exemption.

for a rent v1 the benefits of this (saves account space, simpler) outweigh the cons. I think thats fair because the "fee" goes to the program's rent_sponsor PDA, funding the next user's PDA rent exemption.

* added derive_rent_sponsor!(program_id) macro
* allow rust expressions as seeds to add_compressible_instructions!() macro
* allow seed references to other instruction accounts as well as instruction_data
* made macros modular, moved processors to token-sdk and sdk (compress_runtime.rs, decompress_runtime.rs)

added tests for the macros at every level:
* sdk-compressible-test is a manual implementation of the hooks
* csdk-anchor-derived-test uses most smaller macros
* csdk-anchor-full-derived-test uses add_compressible_instructions macro

## Summary by CodeRabbit

* **New Features**
  * New compressible macros/derives and runtime processors for idempotent compression/decompression (PDAs & tokens).
  * RentSponsor type and rent-sponsor PDA derivation helpers.

* **Documentation**
  * Added comprehensive Compressible GUIDE and README with PDA-only, token, cToken, cMint, and client workflows.

* **Improvements**
  * Config extended with compression authority, rent settings, and write top-up; simplified per-account idempotent compression flow and safer edge-case handling (fewer required parameters).

<sub>✏️ Tip: You can customize this high-level summary in your review settings.</sub>

### CODE REVIEW

File: sdk-libs/program-test/Cargo.toml [hunk 1]
```diff
@@ -7,8 +7,18 @@ edition = "2021"
...
+v2 = [
+    "light-client/v2",
+    # needed for auto-compress
+    "dep:account-compression",
+    "dep:light-compressed-token",
+    "dep:light-ctoken-types",
+    "dep:light-compressible",
+    "dep:light-registry",
+    "dep:light-batched-merkle-tree",
+    "dep:light-concurrent-merkle-tree",
```
Comment (ananas-block) [inline sdk-libs/program-test/Cargo.toml]: v2 feature must not depend on anchor programs.
That's why compressible functions are only available in devenv for now.
Please don't change the way the features work. We should add manual instruction builders for registry program instructions in a separate pr then we can make compressible available in non devenv.

File: sdk-libs/program-test/Cargo.toml [hunk 2]
```diff
@@ -20,7 +30,7 @@ light-concurrent-merkle-tree = { workspace = true, optional = true }
 light-hasher = { workspace = true, features = ["poseidon", "sha256", "keccak", "std"] }
 light-ctoken-types = { workspace = true, optional = true }
 light-compressible = { workspace = true, optional = true }
-light-compressed-token-sdk = { workspace = true, optional = true }
+light-compressed-token-sdk = { workspace = true }
```
Comment (ananas-block) [inline sdk-libs/program-test/Cargo.toml:23]: Make this optional again else we cannot publish a new version of `light-program-test` until we published `light-compressed-token-sdk`.

File: sdk-libs/macros/src/compressible/GUIDE.md [hunk 3]
```diff
@@ -0,0 +1,199 @@
...
+
+```rust
+use light_sdk::compressible::CompressionInfo;
+use light_sdk_macros::{account, Compressible};
+
+#[account]
+#[derive(Compressible)]
+pub struct UserRecord {
+    #[skip]
+    pub compression_info: Option<CompressionInfo>,
```
Comment (ananas-block) [inline sdk-libs/macros/src/compressible/GUIDE.md]: `#[skip]` doesn't work with the sha optimized `LightAccount` implementation

File: sdk-libs/compressible-client/src/lib.rs [hunk 4]
```diff
@@ -63,12 +71,10 @@ pub mod compressible_instruction {
     /// SHA256("global:decompress_accounts_idempotent")[..8]
     pub const DECOMPRESS_ACCOUNTS_IDEMPOTENT_DISCRIMINATOR: [u8; 8] =
         [114, 67, 61, 123, 234, 31, 1, 112];
-    /// SHA256("global:compress_token_account_ctoken_signer")[..8]
-    pub const COMPRESS_TOKEN_ACCOUNT_CTOKEN_SIGNER_DISCRIMINATOR: [u8; 8] =
-        [243, 154, 172, 243, 44, 214, 139, 73];
     /// SHA256("global:compress_accounts_idempotent")[..8]
     pub const COMPRESS_ACCOUNTS_IDEMPOTENT_DISCRIMINATOR: [u8; 8] =
-        [89, 130, 165, 88, 12, 207, 178, 185];
+        // [89, 130, 165, 88, 12, 207, 178, 185];
```
Comment (ananas-block) [inline sdk-libs/compressible-client/src/lib.rs]: leftover?

File: sdk-libs/program-test/Cargo.toml [hunk 5]
```diff
@@ -20,7 +19,7 @@ light-concurrent-merkle-tree = { workspace = true, optional = true }
 light-hasher = { workspace = true, features = ["poseidon", "sha256", "keccak", "std"] }
 light-ctoken-types = { workspace = true, optional = true }
 light-compressible = { workspace = true, optional = true }
-light-compressed-token-sdk = { workspace = true }
+light-compressed-token-sdk = { workspace = true, optional = true }
```
Comment (ananas-block) [inline sdk-libs/program-test/Cargo.toml]: I think this is the reason why the ci tests are failing

File: sdk-libs/program-test/src/program_test/config.rs [hunk 6]
```diff
@@ -68,7 +68,7 @@ impl ProgramTestConfig {
         }
     }
 
-    #[cfg(feature = "v2")]
+    #[cfg(feature = "devenv")]
```
Comment (ananas-block) [inline sdk-libs/program-test/src/program_test/config.rs]:
```suggestion
    #[cfg(feature = "v2")]
```

File: sdk-libs/program-test/src/indexer/test_indexer.rs [hunk 7]
```diff
@@ -870,9 +870,9 @@ impl Indexer for TestIndexer {
         _merkle_tree_pubkey: [u8; 32],
         _config: Option<IndexerRpcConfig>,
     ) -> Result<Response<Items<[u8; 32]>>, IndexerError> {
-        #[cfg(not(feature = "v2"))]
+        #[cfg(not(feature = "devenv"))]
         unimplemented!("get_subtrees");
-        #[cfg(feature = "v2")]
+        #[cfg(feature = "devenv")]
```
Comment (ananas-block) [inline sdk-libs/program-test/src/indexer/test_indexer.rs]: pls revert

File: sdk-libs/program-test/src/indexer/test_indexer.rs [hunk 8]
```diff
@@ -17,7 +17,7 @@ use async_trait::async_trait;
 use borsh::BorshDeserialize;
 #[cfg(feature = "devenv")]
 use light_batched_merkle_tree::merkle_tree::BatchedMerkleTreeAccount;
-#[cfg(feature = "v2")]
+#[cfg(feature = "devenv")]
```
Comment (ananas-block) [inline sdk-libs/program-test/src/indexer/test_indexer.rs]:
```suggestion
#[cfg(feature = "v2")]
```

File: sdk-libs/macros/src/rent_sponsor.rs [hunk 9]
```diff
@@ -0,0 +1,166 @@
...
+            return syn::Error::new(
+                proc_macro2::Span::call_site(),
+                "Invalid program ID format. Expected a base58 encoded public key",
+            )
+            .to_compile_error()
+            .into();
+        }
+    };
+
+    let seeds = &[b"rent_sponsor".as_slice(), &version_u16.to_le_bytes()[..]];
```
Comment (ananas-block) [inline sdk-libs/macros/src/rent_sponsor.rs]: can we use a method for the derivation?

File: sdk-libs/macros/src/rent_sponsor.rs [hunk 10]
```diff
@@ -0,0 +1,166 @@
...
+    let (pda, bump) = solana_pubkey::Pubkey::find_program_address(seeds, &program_id);
```
Comment (ananas-block) [inline sdk-libs/macros/src/rent_sponsor.rs]: can we use a method for the derivation?

File: sdk-libs/macros/src/rent_sponsor.rs [hunk 11]
```diff
@@ -0,0 +1,166 @@
...
+
+    let pda_bytes = pda.to_bytes();
+    let bytes = pda_bytes
+        .iter()
+        .map(|b| proc_macro2::Literal::u8_unsuffixed(*b));
```
Comment (ananas-block) [inline sdk-libs/macros/src/rent_sponsor.rs:118]: we could add a helper function that returns `[#(#program_id_literals),*]` this is used in multiple places

File: sdk-libs/macros/src/lib.rs [hunk 12]
```diff
@@ -124,109 +121,89 @@ pub fn light_accounts_derive(input: TokenStream) -> TokenStream {
...
+//     let input = parse_macro_input!(input as ItemStruct);
+//     discriminator_sha(input)
+//         .unwrap_or_else(|err| err.to_compile_error())
+//         .into()
+// }
+
 /// Makes the annotated struct hashable by implementing the following traits:
 ///
-/// - [`ToByteArray`](light_hasher::to_byte_array::ToByteArray), which makes the struct
+/// - [`AsByteVec`](light_hasher::bytes::AsByteVec), which makes the struct
```
Comment (ananas-block) [inline sdk-libs/macros/src/lib.rs]: pls revert `AsByteVec` doesn't exist anymore

File: sdk-libs/macros/src/lib.rs [hunk 13]
```diff
@@ -236,47 +213,373 @@ pub fn light_hasher(input: TokenStream) -> TokenStream {
...
+/// }
+/// ```
+///
+/// ## Example - Custom Compression
+///
+/// ```ignore
+/// #[derive(Compressible)]
+/// #[compress_as(start_time = 0, end_time = None, score = 0)]
+/// pub struct GameSession {
+///     #[skip]
```
Comment (ananas-block) [inline sdk-libs/macros/src/lib.rs]: `#[skip]` is not applied if we use `sha256::LightAccount` (the default `LightAccount`)

File: sdk-libs/macros/src/lib.rs [hunk 14]
```diff
@@ -236,47 +213,373 @@ pub fn light_hasher(input: TokenStream) -> TokenStream {
...
+/// ```ignore
+/// use light_sdk::derive_light_cpi_signer;
+///
+/// pub const LIGHT_CPI_SIGNER: CpiSigner =
+///     derive_light_cpi_signer!("8Ld9pGkCNfU6A7KdKe1YrTNYJWKMCFqVHqmUvjNmER7B");
+/// ```
+#[proc_macro]
+pub fn derive_light_cpi_signer(input: TokenStream) -> TokenStream {
+    cpi_signer::derive_light_cpi_signer(input)
+}
```
Comment (ananas-block) [inline sdk-libs/macros/src/lib.rs]: pls remove I moved this to `program-libs/macros` since it is used in the `ctoken` program

File: sdk-libs/macros/src/lib.rs [hunk 15]
```diff
@@ -236,47 +213,373 @@ pub fn light_hasher(input: TokenStream) -> TokenStream {
...
+    rent_sponsor::derive_light_rent_sponsor_pda(input)
 }
 
+/// Derives a complete Rent Sponsor configuration for a program at compile time.
+///
+/// Returns ::light_sdk_types::RentSponsor { program_id, rent_sponsor, bump, version }.
+///
+/// ## Example
+///
+/// ```ignore
```
Comment (ananas-block) [inline sdk-libs/macros/src/lib.rs:539]:
```suggestion
/// ```rust
```

File: sdk-libs/macros/src/lib.rs [hunk 16]
```diff
@@ -236,47 +213,373 @@ pub fn light_hasher(input: TokenStream) -> TokenStream {
...
+pub fn derive_light_cpi_signer(input: TokenStream) -> TokenStream {
+    cpi_signer::derive_light_cpi_signer(input)
+}
+/// Derives a Rent Sponsor PDA for a program at compile time.
+///
+/// Seeds: ["rent_sponsor", <u16 version little-endian>]
+///
+/// ## Example
+///
+/// ```ignore
```
Comment (ananas-block) [inline sdk-libs/macros/src/lib.rs:522]:
```suggestion
/// ```rust
```

File: programs/system/Cargo.toml [hunk 17]
```diff
@@ -25,8 +25,8 @@ reinit = []
 default = ["reinit"]
 test-sbf = []
 readonly = []
-profile-program = ["light-program-profiler/profile-program"]
-profile-heap = ["light-program-profiler/profile-heap", "dep:light-heap"]
+profile-program = []
+profile-heap = ["dep:light-heap"]
```
Comment (ananas-block) [inline programs/system/Cargo.toml]: pls revert

File: sdk-libs/compressed-token-sdk/src/decompress_runtime.rs [hunk 18]
```diff
@@ -0,0 +1,160 @@
...
+    post_system_accounts: &[AccountInfo<'info>],
+    has_pdas: bool,
+    program_id: &Pubkey,
+) -> Result<(), ProgramError>
+where
+    V: CTokenSeedProvider<Accounts<'info> = A>,
+    A: 'info,
+{
+    let mut token_decompress_indices: Box<Vec<crate::instructions::DecompressFullIndices>> =
+        Box::new(Vec::with_capacity(ctoken_accounts.len()));
```
Comment (ananas-block) [inline sdk-libs/compressed-token-sdk/src/decompress_runtime.rs]: `Vec` is already heap memory there is no point to `Box` it as well.

Comment (ananas-block) [reply sdk-libs/compressed-token-sdk/src/decompress_runtime.rs]:
```suggestion
    let mut token_decompress_indices: Vec<crate::instructions::DecompressFullIndices> =
        Vec::with_capacity(ctoken_accounts.len());
```

File: sdk-libs/compressed-token-sdk/src/lib.rs [hunk 19]
```diff
@@ -13,7 +14,7 @@ pub mod utils;
 use anchor_lang::{AnchorDeserialize, AnchorSerialize};
 #[cfg(not(feature = "anchor"))]
 use borsh::{BorshDeserialize as AnchorDeserialize, BorshSerialize as AnchorSerialize};
-// Re-export
+pub use decompress_runtime::{process_decompress_tokens_runtime, CTokenSeedProvider};
```
Comment (ananas-block) [inline sdk-libs/compressed-token-sdk/src/lib.rs]: why reexport?

File: sdk-libs/compressed-token-sdk/src/compressible/decompress_runtime.rs [hunk 20]
```diff
@@ -0,0 +1,160 @@
...
+        if derived_token_account_address != *owner_info.key {
+            msg!(
+                "derived_token_account_address: {:?}",
+                derived_token_account_address
+            );
+            msg!("owner_info.key: {:?}", owner_info.key);
+            return Err(ProgramError::InvalidAccountData);
+        }
+
+        let seed_refs: Vec<&[u8]> = ctoken_signer_seeds.iter().map(|s| s.as_slice()).collect();
```
Comment (ananas-block) [inline sdk-libs/compressed-token-sdk/src/compressible/decompress_runtime.rs:121]: ~Since arrays of references have a small memory footprint `ArrayVec` with capacity 16 should not exceed stack bounds and be more CU efficient.~ 
Really interesting that `Vec` is 3x more efficient in the general case [see benchmarks](https://github.com/Lightprotocol/cu-library?tab=readme-ov-file#95-seed-references).

File: sdk-libs/program-test/src/accounts/initialize.rs [hunk 21]
```diff
@@ -178,7 +178,7 @@ pub async fn initialize_accounts<R: Rpc + TestRpc>(
         )
         .await?;
     }
-    #[cfg(feature = "v2")]
+    #[cfg(feature = "devenv")]
```
Comment (ananas-block) [inline sdk-libs/program-test/src/accounts/initialize.rs]: pls revert

**Comment by SwenSchaeferjohann** [comment]
@coderabbitai reset

**Comment by SwenSchaeferjohann** [comment]
@coderabbitai pause

**Review by ananas-block — APPROVED** [review]
